gossipsub.go

```diff
@@ -570,22 +746,47 @@ func (gs *GossipSubRouter) heartbeat() {
 
 	tograft := make(map[peer.ID][]string)
 	toprune := make(map[peer.ID][]string)
+	noPX := make(map[peer.ID]bool)
 
 	// clean up expired backoffs
 	gs.clearBackoff()
 
 	// maintain the mesh for topics we have joined
 	for topic, peers := range gs.mesh {
+		prunePeer := func(p peer.ID) {
+			gs.tracer.Prune(p, topic)
+			delete(peers, p)
+			gs.untagPeer(p, topic)
+			gs.addBackoff(p, topic)
+			topics := toprune[p]
+			toprune[p] = append(topics, topic)
+		}
+
+		// compute mesh peer scores
+		scores := make(map[peer.ID]float64)
+		for p := range peers {
+			scores[p] = gs.score.Score(p)
+		}
+
+		// drop all peers with negative score, without PX
+		for p := range peers {
+			if scores[p] < 0 {
+				log.Debugf("HEARTBEAT: Prune peer %s with negative score [score = %f, topic = %s]", p, scores[p], topic)
+				prunePeer(p)
+				noPX[p] = true
+			}
+		}
 
 		// do we have enough peers?
 		if len(peers) < GossipSubDlo {
 			backoff := gs.backoff[topic]
 			ineed := GossipSubD - len(peers)
 			plst := gs.getPeers(topic, ineed, func(p peer.ID) bool {
-				// filter our current peers and peers we are backing off
+				// filter our current peers, peers we are backing off, and peers with negative score
 				_, inMesh := peers[p]
 				_, doBackoff := backoff[p]
-				return !inMesh && !doBackoff
+				score := gs.score.Score(p)
```

Comment:
Could we reuse the already computed scores in `scores` here?

Reply:
no, we have scores for peers _in_ the mesh, not the other ones. But the code should short-circuit and avoid calling `Score` if `inMesh` or `doBackoff` is true.

Reply:
I short-circuited the score calculation to only occur if `!inMesh && !doBackoff`.